gtsfm/retriever/netvlad_retriever.py

```diff
@@ -0,0 +1,175 @@
+        """Compute a similarity matrix between all pairs of images.
+
+        We use block matching, to avoid excessive memory usage.
+        We cannot fit more than 50x50 sized block into memory, on a 16 GB RAM machine.
+
+        A similar blocked exhaustive matching implementation can be found in COLMAP:
+        https://github.com/colmap/colmap/blob/dev/src/feature/matching.cc#L899
+
+        Returns:
+            sim: tensor of shape (num_images, num_images) representing similarity matrix.
+        """
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        if num_images > MAX_NUM_IMAGES:
+            raise RuntimeError("Cannot construct similarity matrix of this size.")
+
+        sim = torch.zeros((num_images, num_images))
+        num_blocks = math.ceil(num_images / self._blocksize)
+
+        for block_i in range(num_blocks):
+            for block_j in range(num_blocks):
+                # only compute the upper triangular portion of the similarity matrix.
+                if block_i > block_j:
+                    continue
+                logger.info("Computing matching block (%d/%d,%d/%d)", block_i, num_blocks - 1, block_j, num_blocks - 1)
+                block_i_query_descs = []
+                block_j_query_descs = []
+
+                i_start = block_i * self._blocksize
+                i_end = (block_i + 1) * self._blocksize
+                i_end = min(i_end, num_images)
+
+                j_start = block_j * self._blocksize
+                j_end = (block_j + 1) * self._blocksize
+                j_end = min(j_end, num_images)
+
+                block_i_idxs = np.arange(i_start, i_end)
+                block_j_idxs = np.arange(j_start, j_end)
+
+                for i in block_i_idxs:
+                    image = loader.get_image(i)
+                    block_i_query_descs.append(self._global_descriptor_model.describe(image))
+
+                for j in block_j_idxs:
+                    image = loader.get_image(j)
+                    block_j_query_descs.append(self._global_descriptor_model.describe(image))
+
+                # Form (K,D) for K images.
+                block_i_query_descs = torch.from_numpy(np.array(block_i_query_descs))
+                block_j_query_descs = torch.from_numpy(np.array(block_j_query_descs))
+
+                # Einsum equivalent to (img_descs @ img_descs.T)
+                sim_block = torch.einsum("id,jd->ij", block_i_query_descs.to(device), block_j_query_descs.to(device))
+                sim[i_start:i_end, j_start:j_end] = sim_block
+
+        return sim
+
+
+def pairs_from_score_matrix(
+    scores: torch.Tensor, invalid: np.array, num_select: int, min_score: Optional[float] = None
+) -> List[Tuple[int, int]]:
+    """Identify image pairs from a score matrix.
+
+    Args:
+        scores: (K1,K2) for matching K1 images against K2 images.
+        invalid: (K1,K2) boolean array indicating invalid match pairs (e.g. self-matches).
```

Comment:
Where 1s are invalid and 0s are valid?

Reply:
it's actually a boolean matrix (I cast np.ones() to bool type before passing it in).